### PR TITLE
Clean up CompactionDirectives default impl

### PR DESCRIPTION
**Pull request by milleruntime**

### CODE REVIEW

**Comment by ctubbsii** [comment]
@milleruntime I'm trying to review this, but it's taking time, because I'm finding I have soooo many questions about the compaction stuff, and having a hard time understanding where to begin. I'm sure I could figure some of this out if I spent some more time with it, but my initial thoughts/questions are:

1. What's a compaction directive, vs. a compaction service, vs. a compaction dispatcher? (these appear to be the different categories of things touched by this change)
2. Why was CompactionDirectiveImpl a package-private class stored in the SPI package instead of a separate implementation package?
3. What was the compaction directives builder that is being deleted in this PR even for (it had no javadoc)? Why did it carry a service ID, and what even is a service ID?
4. Does the retention of CompactionDirectiveImpl.setService in your PR imply that CompactionsDirectiveImpl is mutable? Should it be mutable? Is it okay to be mutable?
5. Is the loop in SimpleCompactionDispatcher that calls defaultService.setService(v) a good idea? What's the end goal being pursued here if the service that is being held in the defaultService object is being swapped out repeatedly in a loop?
6. Naming-wise, what does it mean for a "defaultService" to have a "setService" method? What is the relationship between the "service" that the variable name is referring to and the "service" that the method name is referring to?

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectives.java [hunk 1]
```diff
@@ -30,19 +30,4 @@
    * @return The service where a compaction should run.
    */
   CompactionServiceId getService();
-
-  /**
-   * @since 2.1.0
-   */
-  public static interface Builder {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectives.java:37]: This builder was intended to be used by someone writing a CompactionDispatcher.  If someone were to write a compaction dispatcher now they would need to implement CompactionDirective.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java [hunk 2]
```diff
@@ -72,19 +72,17 @@
...
     }
 
     for (CompactionKind ctype : CompactionKind.values()) {
       String service = params.getOptions().get("service." + ctype.name().toLowerCase());
-      if (service == null)
-        services.put(ctype, defaultService);
-      else
-        services.put(ctype, CompactionDirectives.builder().setService(service).build());
+      if (service != null)
+        defaultService.setService(service);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: This is changing a static object used by many different dispatcher instance.  This would cause a config change on one table to affect all other tables.  Also the builder used to create immutable objects.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionsDirectiveImpl.java [hunk 3]
```diff
@@ -18,64 +18,28 @@
...
-
-import org.apache.accumulo.core.spi.compaction.CompactionDirectives.Builder;
-
-import com.google.common.base.Preconditions;
-
 /**
- * This class intentionally package private. This implementation is odd because it supports zero
- * object allocations for {@code CompactionDirectives.builder().build()}.
+ * This class intentionally package private. It supports one object allocation for
+ * {@code CompactionDirectives}.
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionsDirectiveImpl.java:23]:
```suggestion
 * This class intentionally package private.
```

The old code implemented a copy on write strategy to avoid object allocation when building the default.  If that is to be removed I don't think a comment is needed saying objects are allocated.

**Review by keith-turner — CHANGES_REQUESTED** [review]
> What's a compaction directive, vs. a compaction service, vs. a compaction dispatcher? (these appear to be the different categories of things touched by this change)

A compaction dispatcher dispatcher compactions per table to compaction services.  A compactions service runs compactions and there are one or more in a tablet servers.  Compaction directive exists a return type for [CompactionDispatcher.dispatch()](https://github.com/apache/accumulo/blob/7e3c67d0c0b831d7d29c69e24d07dd700f24c5e0/core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDispatcher.java#L90).  The reason CompactionDirective exist is to make CompactionDispatcher easier to evolve.

> Why was CompactionDirectiveImpl a package-private class stored in the SPI package instead of a separate implementation package?

It could have been placed in another package outside of the SPI packages instead of being made package private.  One slight advantage of package private is that it can not be accessed from any other package.   It existed solely to implement the builder.  One disadvantage of making it package private is that someone may inadvertently make it public.

> Does the retention of CompactionDirectiveImpl.setService in your PR imply that CompactionsDirectiveImpl is mutable? Should it be mutable? Is it okay to be mutable?

I think its much cleaner to make it immutable, as that makes code easier to reason about.  The builder that was removed would build an immutable object.

**Comment by keith-turner** [comment]
CompactionDispatcher and CompactionDirective follow the same pattern as [ScanDispatcher](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/spi/scan/ScanDispatcher.java) and [ScanDirective](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/spi/scan/ScanDispatcher.java).  In my opinion it would be good to keep them consistent.  ScanDispatcher used to return a single string for its dispatch method in 2.0.0.  Then in 2.1.0 cache information was added to the scan dispatch method resulting in the need to return an object.  That is why I chose to start off returning an object for the compaction dispatch method.

**Comment by milleruntime** [comment]
> CompactionDispatcher and CompactionDirective follow the same pattern as [ScanDispatcher](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/spi/scan/ScanDispatcher.java) and [ScanDirectives](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/spi/scan/ScanDirectives.java). In my opinion it would be good to keep them consistent. ScanDispatcher used to return a single string for its dispatch method in 2.0.0. Then in 2.1.0 cache information was added to the scan dispatch method resulting in the need to return an object. That is why I chose to start off returning an object for the compaction dispatch method.

There seems to be similar issues with ScanDispatcher, mainly having a builder where only one method should be called to set a single object on another class. My concern is that it seems like a lot of extra code for just calling `new ScanDirectivesImpl().setX()` and the builder returned allows multiple `setX()` calls, which would only create throw away objects.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java (hunk 2 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: It was not clear to me that this class was immutable, since it contained multiple `setter` methods. I think a POJO with final members set in the constructor would be a lot more clear.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectives.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectives.java:37]: If CompactionDirective is for returning information to the user, than it shouldn't need a builder (similar to how we return Environment objects). If the user needs to set objects in it for their implementation, perhaps a separate utility method would be better.

**Comment by keith-turner** [comment]
> There seems to be similar issues with ScanDispatcher, mainly having a builder where only one method should be called to set a single object on another class.

For the ScanDispatcher three things can be set.  In 2.0.0 only one thing could be set for the ScanDispatcher.  For 2.1.0 to additional things were added.  Because ScanDispatcher did not follow the pattern in 2.0.0 that it does now, the addition of new things to return was a painful.  The design of this interface is the way it is to support SPI evolution where new things can be returned in the future.  When writing the ScanDispatcher in 2.0.0 I never could have imagined additionally returning cache directives in 2.1.0.

> My concern is that it seems like a lot of extra code for just calling new ScanDirectivesImpl().setX() and the builder

Someone writing a ScanDispatcher or CompactionDispatcher would never interact with ScanDirectivesImpl or CompactionDirectivesImpl.  Those are internal implementations hidden from users.  When making changes I would suggest to start with the interfaces CompactionDirectives and ScanDirectives and changes those to how you think they should be w/o regard for ScanDirectivesImpl and CompactionDirectivesImpl.  Also considering how a user writing a CompactionDispatcher would use CompactionDirectives, as the interfaces only exists to support CompactionDispatcher.

**Comment by ctubbsii** [comment]
Perhaps a stupid question, but if it's designed as complex, extensible return type for the `Dispatcher.dispatch()` method, why not call the returned object a "Dispatch" rather than a "Directive"?

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java (hunk 2 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: Changing CompactionDirectives from an interface w/ a builder to a POJO with a single constructor and final instance vars would be an alternative.  I intentionally chose not to take that route because I have seen too many things in Accumulo end up with lots over overloaded constructors and methods over time.  The builder pattern is a good way to avoid this.  If nothing is ever added the POJO would be better overall, but I like I mentioned with the ScanDirectives I never could have imagined adding the caching directives in 2.1.0 beforehand.     Also I don't think there is much complexity currently from the perspective of someone using the current incarnation of CompactionDirectives.  There is certainly a lot of complexity in the impl class and some of that comes from trying to avoid object allocations.

**Comment by keith-turner** [comment]
> Perhaps a stupid question, but if it's designed as complex, extensible return type for the Dispatcher.dispatch() method, why not call the returned object a "Dispatch" rather than a "Directive"?

That name seems fine w/ me.  If changed I think it would make sense to consider ScanDispatcher also.  At this point I would personally prioritize consistency between the two.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java (hunk 2 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: Another alternative to the POJO would be an interface w/ static methods that create an instance.  This has the same potential problem as constructors, there could be lots of overloaded methods in the future.  I think one nice thing about an interface w/ a static method is that if in the future multiple things are going to be added then a static builder() method could be added.  

Something like the following in 2.1.0

```java
public interface CompactionDirectives {

  /**
   * @return The service where a compaction should run.
   */
  CompactionServiceId getService();
 
  static CompactionDirectives from(CompactionServiceId){...}
}
```

Then in 2.2.0 we think of something else we want to return.

```java
public interface CompactionDirectives {

  /**
   * @return The service where a compaction should run.
   */
  CompactionServiceId getService();
 
  static CompactionDirectives from(CompactionServiceId){}
 /**
  * @since 2.2.0
  */
  TypeX getX();

 /**
  * @since 2.2.0
  */
 static CompactionDirectives from(CompactionServiceId csid, TypeX x){}
}
```

Then in 2.3.0 we think of 5 things we want to add and at that point just add a builder.

```java
public interface CompactionDirectives {

  /**
   * @return The service where a compaction should run.
   */
  CompactionServiceId getService();
 
  static CompactionDirectives from(CompactionServiceId){}
 /**
  * @since 2.2.0
  */
  TypeX getX();

 /**
  * @since 2.2.0
  */
 static CompactionDirectives from(CompactionServiceId csid, TypeX x){}

 /**
  * @since 2.3.0
  */
  interface Builder {...}

 /**
  * @since 2.3.0
  */
  static Builder builder(); 
}
```

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: I am just trying to figure out if there is a way to make your highly evolved API code easier to understand for new developers. After playing around with it all afternoon, I still don't have a way to do that. I just worry because I _know_ that I have looked at the ScanDispatcher code before and yet, it still took me a few minutes to comprehend it again. Oh well, I digress.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: > I am just trying to figure out if there is a way to make your highly evolved API code easier to understand for new developers.

Easier to understand for an Accumulo user writing a compaction dispatcher or easier to understand for an Accumulo developer working on internal code?  For the latter, CompactionDirectivesImpl is complicated.. I may have gotten carried away trying to avoid object allocations which I think is at the root of the complexity.  For the former, if we want to clean up the user facing SPI it would good to do that before 2.1 is released.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: To simplify CompactionDirectivesImpl w/o changing the current SPI, could create two classes like the followin.

```java
class CompactionDirectiveBuilder implements Builder {

  private CompactionServiceId service;
  
  @Override
  public Builder setService(CompactionServiceId service) {
    this.service = service;
    return this;
  }

  @Override
  public Builder setService(String compactionServiceId) {
    this.service = CompactionServiceId.of(compactionServiceId);
    return this;
  }

  @Override
  public CompactionDirectives build() {
    return new CompactionsDirectiveImpl(service);
  }
  
}

class CompactionsDirectiveImpl implements CompactionDirectives {

  private final CompactionServiceId service;
  
  public CompactionsDirectiveImpl(CompactionServiceId service) {
    this.service = service;
  }

  @Override
  public CompactionServiceId getService() {
    return service;
  }

  @Override
  public String toString() {
    return "service=" + service;
  }
}
```

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:84]: > Easier to understand for an Accumulo user writing a compaction dispatcher or easier to understand for an Accumulo developer working on internal code?

I think both would be good goals. But I was first targeting the latter. 

I think splitting the two implementations is a big help. I will replace the changes on this PR with the two separate classes.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionsDirectiveImpl.java (hunk 3 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionsDirectiveImpl.java:23]: I am going to rename this class to `CompactionDirectivesImpl`. This follows the interface name. It was also the only class with an 's' on Compaction for some reason, possibly a typo.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectives.java [hunk 4]
```diff
@@ -43,6 +43,6 @@
   }
 
   public static Builder builder() {
-    return CompactionsDirectiveImpl.DEFAULT_BUILDER;
+    return CompactionDirectivesBuilder.DEFAULT_BUILDER;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectives.java:46]: If there are two concurrent threads they would both be using the same object.

```suggestion
    return new CompactionDirectivesBuilder();
```

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.spi.compaction;
+
+class CompactionDirectivesBuilder implements CompactionDirectives.Builder {
+
+  static final CompactionDirectives.Builder DEFAULT_BUILDER = new CompactionDirectivesBuilder();
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:23]:
```suggestion
```

**Comment by milleruntime** [comment]
@keith-turner After the last changes, I had to add a way to set the default in a977accf7a82ea5fc5dc6a6d5b90a1f4355642a3

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.spi.compaction;
+
+class CompactionDirectivesBuilder implements CompactionDirectives.Builder {
+
+  private CompactionServiceId service;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:23]:
```suggestion
  private static final CompactionServiceId DEFAULT_SERVICE_ID=CompactionServiceId.of("default");
  private CompactionServiceId service = DEFAULT_SERVICE_ID;
```

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java [hunk 7]
```diff
@@ -72,7 +72,7 @@
   public void init(InitParameters params) {
     services = new EnumMap<>(CompactionKind.class);
 
-    var defaultService = CompactionDirectives.builder().build();
+    var defaultService = CompactionDirectives.builder().setService("default").build();
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]:
```suggestion
    var defaultService = CompactionDirectives.builder().build();
```

**Review by keith-turner — COMMENTED** [review]
Made a few suggestions that result in `CompactionDirectives.builder().build()` returning defaults.

**Review by ctubbsii — APPROVED** [review]
I like these changes. The application of the builder pattern is much more intuitive/readable now.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java (hunk 7 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: Or, could use the name of the `SimpleCompactionDispatcher`?

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java (hunk 6 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:23]: I don't see why we need to have a hard-coded default name at all. Why can't this just always be a required parameter provided by the dispatcher?

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:23]: That is kinda what I was going for by putting the name of "default" set in the `SimpleCompactionDispatcher`

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:23]: > I don't see why we need to have a hard-coded default name at all. Why can't this just always be a required parameter provided by the dispatcher?

My thinking was that when the builder builds a default obj that it reflects the defaults in config.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java (hunk 7 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: The service name is used to get compaction service config and uniquely identify the service within the tserver.

Compaction service properties have the following prefix.
```
tserver.compaction.major.service.<service id>
```

With the default service currently having multiple default props w/ the following prefix.

```
tserver.compaction.major.service.default
```

Using the classname  would not result in very nice default props.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: I am fine with the name. I just thought it made more sense setting the value of "default" in the `SimpleCompactionDispatcher` class because this _is_ the default implementation. Your suggestion puts this logic into the builder, which I don't think should have any of the default business logic.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: I agree my name change suggestion doesn't make sense, and that `default` makes more sense here, since that's the name of the service that `SimpleCompactionDispatcher` is dispatching *to*. So, I agree with @milleruntime more than @keith-turner here, and think that setting it in `SimpleCompactionDispatcher` is more intuitive. I would rather the builder simply have it as a required parameter.

Also, I think a lot of this would be more clear if the Directive were called a Dispatch. Then, it would be more clear that what is being built by the builder is the responsibility of the Dispatcher, because it is the Dispatcher's job to set the name of the service that the Dispatch is being dispatched *to*.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: > Also, I think a lot of this would be more clear if the Directive were called a Dispatch. Then, it would be more clear that what is being built by the builder is the responsibility of the Dispatcher, because it is the Dispatcher's job to set the name of the service that the Dispatch is being dispatched to.

This seems reasonable to me. I agree with @keith-turner though that the name should be consistent with the `ScanDispatcher`, since they follow similar patterns. I didn't want to touch any of the Scan* classes in this PR though, as it was intended to clarify `CompactionDirectives` so I think renaming them both could be done separately.

**Comment by milleruntime** [comment]
@keith-turner I am happy with this PR now if you are OK with the changes.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java (hunk 7 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: There are three additional cases to consider.

 * Future additions of fields to CompactionDirectives would require that the builder be able to default those or require the user to specify them breaking existing code written against the SPI. ScanDirectives currently has three fields and one can do ScanDirectives.builder().build() to create a directives with defaults for all three fields.
 * Users writing their own CompactionDispatcher who want to fall back to the default service would need to use the string `default` in their own code.
 * A user can write   CompactionDirectives.builder().build() and it would compile, but I think would result in an NPE at runtime.  If its going to be required, then it would probably be best to structure the code such that its impossible to produce code that compiles w/o specifying it.

The three cases lead me to support the user not having to specify the default when building a CompactionDirectives

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: I see what you are saying that a call to `CompactionDirectives.builder().build()` should work properly and set the default values. Typically this is fine if the values set in there will work for most purposes. This could be my limited understanding of the new compaction code but I think the problem here is that the default values for `CompactionDirectives` are tied directly to the implementation in `SimpleCompactionDispatcher`. 

Yes, if I wrote my own CustomCompactionDispatcher, it should keep working with future updates to CompactionDirectives interface. But why would my CustomCompactionDispatcher call CompactionDirectives.builder().build() ? That would set the values for SimpleCompactionDispatcher. Wouldn't CustomCompactionDispatcher call CompactionDirectives.builder().setService("customCD").build() ?

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: > his could be my limited understanding of the new compaction code but I think the problem here is that the default values for CompactionDirectives are tied directly to the implementation in SimpleCompactionDispatcher

The default values in this case are more tied to the default in Accumulo's Property.java

> But why would my CustomCompactionDispatcher call CompactionDirectives.builder().build() ?

If the javadoc states that one should not do that, then no one should.  Its just an API/SPI design principal that if an API/SPI can be designed such that user errors are caught at compile time instead of runtime, then compile time would be better if it can be done in a reasonable way.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: > API/SPI can be designed such that user errors are caught at compile time instead of runtime

For sure. I can do that in this PR.

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: > The default values in this case are more tied to the default in Accumulo's Property.java

I had a question about this that perhaps you can clarify: am I using the terminology correctly if I say that there is a **major compaction service** named *default*, and that this **service** named *default*, by default, has 3 **executors** named *small*, *medium*, and *large*? If I have the terminology correct so far, is it correct to say that the only default value from the CompactionDirectives builder that ties to what's in Property.java is the name *default* to refer to the named **service**?

My understanding is that the CompactionDirectives is basically just assignment information to send the Directive (which I would call Dispatch) to the specified service. Is that correct?

If I am understanding all this correctly, then I think we should just make `setService()` a required method call before one is capable of calling `build()`. Since the responsibility of assigning the dispatch to a specific service is pretty much the only thing that the Dispatcher is responsible for doing, then I don't think it should rely on any defaults in the builder... but declare it explicitly.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/compaction/SimpleCompactionDispatcher.java:75]: > that this service named default, by default, has 3 executors named small, medium, and large?

That is correct and the by the service has a planner configured in Property.java

> only default value from the CompactionDirectives builder that ties to what's in Property.java is the name default to refer to the named service?

That is correct.

> If I am understanding all this correctly, then I think we should just make setService() a required method call before one is capable of calling build() (possibly calling it toService()

If it's required at compile time then I would be ok w/ that change.  Also I agreep setService is not a good name for a fluent builder, something else would be better like toService().

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.accumulo.core.spi.compaction;
+
+class CompactionDirectivesBuilder
+    implements CompactionDirectives.Builder, CompactionDirectives.ServiceBuilder {
+
+  private CompactionServiceId service;
+
+  @Override
+  public CompactionDirectives.Builder setService(CompactionServiceId service) {
+    this.service = service;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:28]:
```suggestion
    this.service = Objects.requireNonNull(service);
```

**Review by keith-turner — APPROVED** [review]
I think it would be good to rename setService as @ctubbsii  had suggested.

File: core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.spi.compaction;
+
+class CompactionDirectivesBuilder
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/compaction/CompactionDirectivesBuilder.java:21]: May be good to add note/warning that the class is intentionally package private.

**Comment by milleruntime** [comment]
> I think it would be good to rename setService as @ctubbsii had suggested.

I am not sure what you are referring to. Sorry its hard to follow with all the comments.

**Comment by milleruntime** [comment]
> > I think it would be good to rename setService as @ctubbsii had suggested.
> 
> I am not sure what you are referring to. Sorry its hard to follow with all the comments.

> If it's required at compile time then I would be ok w/ that change. Also I agreep setService is not a good name for a fluent builder, something else would be better like toService().

Nevermind found it. Sorry I missed your response.
